### ISSUE TITLE
Move -Wunaligned-access to W=1

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1569]
A new clang warning was introduced in https://github.com/llvm/llvm-project/commit/35737df4dcd28534bd3090157c224c19b501278a  and [the Linaro TCWG CI reported](https://lore.kernel.org/r/1454838115.14607.1642854713913@jenkins.jenkins/) that it breaks `ARCH=arm allmodconfig` due to `-Werror`:

```
# First few build errors in logs:
# 00:03:03 ./include/linux/tpm.h:288:2: error: field  within 'struct tpm_header' is less aligned than 'union tpm_header::(anonymous at ./include/linux/tpm.h:288:2)' and is usually due to 'struct tpm_header' being packed, which can lead to unaligned accesses [-Werror,-Wunaligned-access]
# 00:03:03 make[3]: *** [scripts/Makefile.build:277: security/keys/encrypted-keys/encrypted.o] Error 1
# 00:03:03 make[2]: *** [scripts/Makefile.build:540: security/keys/encrypted-keys] Error 2
# 00:03:05 ./include/linux/tpm.h:288:2: error: field  within 'struct tpm_header' is less aligned than 'union tpm_header::(anonymous at ./include/linux/tpm.h:288:2)' and is usually due to 'struct tpm_header' being packed, which can lead to unaligned accesses [-Werror,-Wunaligned-access]
# 00:03:05 make[3]: *** [scripts/Makefile.build:277: security/keys/trusted-keys/trusted_core.o] Error 1
# 00:03:06 ./include/linux/tpm.h:288:2: error: field  within 'struct tpm_header' is less aligned than 'union tpm_header::(anonymous at ./include/linux/tpm.h:288:2)' and is usually due to 'struct tpm_header' being packed, which can lead to unaligned accesses [-Werror,-Wunaligned-access]
# 00:03:06 make[3]: *** [scripts/Makefile.build:277: security/keys/trusted-keys/trusted_tpm1.o] Error 1
# 00:03:07 ./include/linux/tpm.h:288:2: error: field  within 'struct tpm_header' is less aligned than 'union tpm_header::(anonymous at ./include/linux/tpm.h:288:2)' and is usually due to 'struct tpm_header' being packed, which can lead to unaligned accesses [-Werror,-Wunaligned-access]
# 00:03:07 make[3]: *** [scripts/Makefile.build:277: security/keys/trusted-keys/trusted_tpm2.o] Error 1
# 00:03:08 ./include/linux/tpm.h:288:2: error: field  within 'struct tpm_header' is less aligned than 'union tpm_header::(anonymous at ./include/linux/tpm.h:288:2)' and is usually due to 'struct tpm_header' being packed, which can lead to unaligned accesses [-Werror,-Wunaligned-access]
```

I disabled `CONFIG_WERROR` and built `ARCH=arm allmodconfig`, which reveals this warning triggers a lot in kernel code: https://gist.github.com/58ce543c7a8792e4c0849a81d0ef467d

It looks like there are 70 unique instances of the warning but it shows up almost 1300 times because the majority of warnings are in header files:

```
$ rg -N -- -Wunaligned-access build.log &| sort &| uniq &| wc -l
70

$ rg -c -- -Wunaligned-access build.log
1284
```

I suspect this warning might need to be disabled in the same vein as `-Waddress-of-packed-member`, since I assume that we would have seen this issue in practice by now on real hardware.

cc @arndb, as you will hit this with a newer clang while doing randconfig testing.

**Comment by arndb** [comment]
Thanks for letting me know about it in advance. I suppose this may be worth fixing globally as it is less common than the -Wunaligned-access warning but equally likely to cause actual problems. The architectures that would be hit by this appear to all work around unaligned accesses using exception fixup handlers, which make any code path that hits it really slow, and having a way for the compiler to statically warn about it would be a great help.

**Comment by nickdesaulniers** [comment]
Should we defer this to W=1 builds for now?

**Comment by nathanchance** [comment]
The comments in `scripts/Makefile.extrawarn`:

```
# W=1 - warnings which may be relevant and do not occur too often

# W=2 - warnings which occur quite often but may still be relevant
```

We probably want this in `W=2` for now, given how noisy it is in an `ARCH=arm allmodconfig` build.

However, I don't think we want to do the traditional `-Wno-unaligned-access` in the `W=1` and no `W=1` block then `-Wunaligned-access` in the `W2` block because this warning is only enabled under certain circumstances within the frontend (see all the `CmdArgs.push_back("-Wunaligned-access")` calls in the commit that added the warning), rather that universally via `-Wall` or something. If we add `-Wunaligned-access` to `W=2`, all architectures will show that warning, rather than just `ARCH=arm` and `ARCH=arm64`, which is what will happen currently.

I'll think more tomorrow about how to work around this (aside from bringing `CONFIG_ARM` and `CONFIG_ARM64` into it).

**Comment by nathanchance** [comment]
Patch submitted: https://lore.kernel.org/r/20220201232229.2992968-1-nathan@kernel.org/

**Comment by nathanchance** [comment]
v2: https://lore.kernel.org/r/20220202230515.2931333-1-nathan@kernel.org/

I made this issue about moving the warning to `W=1`, which will be closed once the above patch lands in mainline. I opened https://github.com/ClangBuiltLinux/linux/issues/1576, which will track turning it back on.

**Comment by nathanchance** [comment]
Patch accepted: https://git.kernel.org/masahiroy/linux-kbuild/c/1cf5f151d25fcca94689efd91afa0253621fb33a

**Comment by nathanchance** [comment]
Merged into mainline: https://git.kernel.org/linus/1cf5f151d25fcca94689efd91afa0253621fb33a

It will go back to stable automatically, I'll make sure to watch for the FAILED emails in case it does not go back as far as it needs to go.
